rfcs/000-search-revamp.md

```diff
@@ -0,0 +1,124 @@
+technical state of search has not changed much from MVP.
+
+We see search as serving two distinct (but related) use cases:
+  - known-data searches, which have a starting point (e.g. tables with a `user_id` column in the `production` schema) 
+    and are more interested in listing the available data matching a known condition
+  - discovery searches, which are looking to find previously-unknown data that would be relevant for solving
+    one's problem
+
+Although we believe Amundsen has been particularly lacking in the discovery search category, our proposal is seeking
+to improve the experience of both search types. 
+
+
+## UI/UX Details
+
+Although the bulk of proposed changes will be to search result quality and relevance, we're also proposing
+a handful of UI/UX changes to filters. We're currently working on designs and will share those as soon as they're
+available.
+
+#### Multiple Values Per Filter.
+Allow multiple values (likely comma-separated) for filters, which has been very highly requested. 
+
+There is an obvious question of whether multiple values would be AND'ed or OR'ed together. Our goal is to provide
+additional expressiveness without going down the path of supporting full logical expressions. To better understand the problem,
+keep in mind that there's two kind of fields: single-valued fields (table name, schema name) and multi-valued fields (columns, tags, badges).
+Our proposed tradeoff between expressiveness and simplicity is to combine filter values for single-valued 
+fields with OR (e.g. "staging,production" filter applied to the "schema" field)
+whereas for multi-valued fields we'd give users an options to choose AND or OR 
+(a user would choose AND/OR to apply the filters "product_id,user_id" to the "column" field)
+
+#### Negation filtering
+Another highly requested feature is specifying negative filters - e.g. "schema is not temp, tmp, or temporary".
+As above, we want to avoid going down the rabbit hole of full-blown logical expressions. Likely, we'd implement
+this on the frontend as a separate set of filter boxes below the current set. Stay tuned for designs. 
+
+#### Type-ahead suggestions for filters
+Because values entered as filters are not interpreted or analyzed (this is the current behavior and is not expected to change),
+it can be hard for users to remember/guess the exact values to filter for. We'd like to provide typeahead
+suggestions that will suggest the most likely filter value given a prefix.
+
+
+## Backend Changes
+As part of this work, most of the elasticsearch [proxy](https://github.com/amundsen-io/amundsen/blob/main/search/search_service/proxy/elasticsearch.py)
+would be re-written, while keeping the API of the search service the same (in part to allow Atlas to stay unchanged).
+
+#### Upgrade Elasticsearch to v7
+Elasticsearch queries/mappings are generally written to a specific version, since a number of features are typically
```

Comment:
Thank you for pointing this out. You're right I was not too explicit in the RFC, in part to give ourselves more space to experiment. There's two features of v7 in particular that excite us, although we're also motivated by just wanting to move to the latest-and-greatest since we'll be touching so much code anyway.

By far the most exciting feature (which should probably have gotten treatment in the RFC - I left it out because we may want to try this internally before pushing to OSS) is [rank_features](https://www.elastic.co/guide/en/elasticsearch/reference/current/rank-features.html). This [blog](https://www.elastic.co/blog/easier-relevance-tuning-elasticsearch-7-0) gives an excellent overview, and also mentions how `function_score`, which is what we use now, is being replaced by `script_score` for those who need full control (which I don't believe we do). 

What we've hypothesized is that one of our biggest issues with search relevance is lack of good usage metrics that would promote high use tables. Usage is very multifaceted (number of dags, number of daily active users, number of downstreams, number of total queries queries against a table, etc etc) and these usages cannot be simply added together. Right now Amundsen has one `total_usage` field and it does not capture these different usages well.

Also - and this is further out - but in my reading of the `rank_features` docs, it seems that the explicit ability to index a sparse metrics opens the door to an MVP of personalized search. If we can index per-user usage into each document, then personalized search becomes a matter of 1 extra line in the query to take `usages.dkunitskiy` into account, alongside `usages.num_downstreams`, `usages.daily_active_users` etc. 

The other v7 addition is the [revamped](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/search-as-you-type.html) search-as-you-type suggestions, which are significantly different in v7. I'm not an expert on this, but we plan to follow the docs to make the filterable fields `search_as_you_type` fields to get filter completions. 

In general you're right that v6 is not deprecated and in my understanding, staying on older ES versions is quite normal. And we also are hesitant to make anyone undergo an ES migration. So we are looking for input on the best approach here, ideally balancing short and long term pain.